django_prbac/models.py

```diff
@@ -65,28 +74,101 @@ class Role(ValidatingModel, models.Model):
     # Methods
     # -------
 
+    @classmethod
+    def get_cache(cls):
+        try:
+            cache = cls.cache
+        except AttributeError:
+            timeout = getattr(settings, 'DJANGO_PRBAC_CACHE_TIMEOUT', 60)
+            cache = cls.cache = DictCache(timeout)
+        return cache
+
+    @classmethod
+    def update_cache(cls):
+        roles = cls.objects.prefetch_related('memberships_granted').all()
+        roles = {role.id: role for role in roles}
+        for role in roles.values():
+            role._granted_privileges = privileges = []
+            # Prevent extra queries by manually linking grants and roles
+            # because Django 1.6 isn't smart enough to do this for us
+            for membership in role.memberships_granted.all():
+                membership.to_role = roles[membership.to_role_id]
+                membership.from_role = roles[membership.from_role_id]
+                privileges.append(membership.instantiated_to_role({}))
+        cache = cls.get_cache()
+        cache.set(cls.ROLES_BY_ID, roles)
+        cache.set(cls.PRIVILEGES_BY_SLUG,
+            {role.slug: role.instantiate({}) for role in roles.values()})
+
+    @classmethod
+    def get_privilege(cls, slug, assignment=None):
+        """
+        Optimized lookup of privilege by slug
+
+        This optimization is specifically geared toward cases where
+        `assignments` is empty.
+        """
+        cache = cls.get_cache()
+        privileges = cache.get(cls.PRIVILEGES_BY_SLUG)
+        if privileges is None:
+            cls.update_cache()
+            privileges = cache.get(cls.PRIVILEGES_BY_SLUG)
+        privilege = privileges.get(slug)
+        if privilege is None:
+            return None
+        if assignment:
+            return privilege.role.instantiate(assignment)
+        return privilege
+
+    def get_cached_role(self):
+        """
+        Optimized lookup of role by id
+        """
+        cache = self.get_cache()
+        roles = cache.get(self.ROLES_BY_ID)
+        if roles is None or self.id not in roles:
+            self.update_cache()
+            roles = cache.get(self.ROLES_BY_ID)
+        return roles[self.id]
```

Comment:
The only case I can think of where it might raise a `KeyError` is if `self` is a new/unsaved role, which come to think of it is a [valid scenario](https://github.com/dimagi/commcare-hq/blob/abcd219fe4d88e974e971220b7da244bc98ec0d1/corehq/apps/domain/middleware.py#L46).